### PR TITLE
Accept final in instanceof pattern

### PR DESCRIPTION
**Pull request by PPazderski**
Javaparser does not accept final in an instanceof pattern but from my understanding of JLS it should be possible.
The following snippet is rejected by javaparser with an unexpected token "final" error but the same code can be compiled with javac and Eclipse compiler.

```
Object o = null;
if (o instanceof final String str) {
   // use string
}
```

### CODE REVIEW

File: javaparser-core/src/main/javacc/java.jj [hunk 1]
```diff
@@ -3278,13 +3278,15 @@ Expression EqualityExpression():
...
 PatternExpr PatternExpression():
 {
+    ModifierHolder modifier;
     ReferenceType type;
     SimpleName name;
 }
 {
-    type = AnnotatedReferenceType()
+    modifier = Modifiers()
+    type = ReferenceType(modifier.annotations)
```
Comment (jlerbsc) [inline javaparser-core/src/main/javacc/java.jj:3287]: Why did you change the type?

Comment (jlerbsc) [reply javaparser-core/src/main/javacc/java.jj:3287]: Looks like a good job to me. Maybe @MysterAitch will take a look at this PR.

Comment (PPazderski) [reply javaparser-core/src/main/javacc/java.jj:3287]: At first, because `InstanceOfExprTest.annotationsOnTheType_patternExpression` failed otherwise. From my understanding AnnotatedReferenceType is just a helper to parse annotation and then create a ReferenceType with those annotations.
```
ReferenceType AnnotatedReferenceType():
{
    NodeList<AnnotationExpr> annotations;
    ReferenceType type;
}
{
    annotations = Annotations()
    type = ReferenceType(annotations)
    { return type; }
}
```
But since `Modifiers()` also parses annotations the `Annotations()` in the later `AnnotatedReferenceType()` couldn't find anything because annotations where already parsed by `Modifiers()`. That's why I took the annotations from the `ModifierHolder` instead.

Comment (jlerbsc) [reply javaparser-core/src/main/javacc/java.jj:3287]: For me it's good. Good work. I'll give @MysterAitch a chance to look at your proposal as well.

**Comment by jlerbsc** [comment]
Thank you for your contribution
